File: benchmarks/r/5_csv_wrangle.R
# Benchmark 5: CSV Wrangling
suppressPackageStartupMessages(library(dplyr))

# Ensure working directory is the benchmarks root
# When run via run_all.ps1, cwd is already correct; this is a safety net
script_dir <- tryCatch(dirname(sys.frame(1)$ofile), error = function(e) NULL)
if (!is.null(script_dir) && nzchar(script_dir)) {
  setwd(file.path(script_dir, ".."))
}

tryCatch({

samples <- read.csv("data/samples.csv", stringsAsFactors = FALSE)
metadata <- read.csv("data/metadata.csv", stringsAsFactors = FALSE)

# Join
joined <- inner_join(samples, metadata, by = "sample_id")

# Group by cohort, summarize
summary_df <- joined %>%
  group_by(cohort) %>%
  summarize(
    count = n(),
    mean_depth = mean(depth),
    mean_quality = mean(quality),
    total_reads = sum(read_count),
    .groups = "drop"
  ) %>%
  arrange(desc(mean_depth))

cat("Cohort Summary:\n")
for (i in seq_len(nrow(summary_df))) {
  row <- summary_df[i, ]
  cat(sprintf("  %s: n=%.0f, depth=%.1f, qual=%.1f, reads=%.0f\n",
              row$cohort, row$count, row$mean_depth, row$mean_quality, row$total_reads))
}

# High-quality filter
hq <- joined %>% filter(quality >= 30, depth >= 20)
cat(sprintf("\nHigh-quality samples: %.0f / %.0f\n", nrow(hq), nrow(joined)))

}, error = function(e) {
  cat(paste("ERROR:", conditionMessage(e), "\n"), file = stderr())
  quit(status = 1, save = "no")
})

File: benchmarks/language/wrangling/python/5_csv_wrangle.py
"""Benchmark 5: CSV Wrangling"""
import csv
from collections import defaultdict
from statistics import mean

# Read samples
with open("data/samples.csv") as f:
    samples = list(csv.DictReader(f))
for s in samples:
    s["depth"] = float(s["depth"])
    s["quality"] = float(s["quality"])
    s["read_count"] = int(s["read_count"])

# Read metadata
with open("data/metadata.csv") as f:
    metadata = {row["sample_id"]: row for row in csv.DictReader(f)}

# Join
joined = []
for s in samples:
    if s["sample_id"] in metadata:
        merged = {**s, **metadata[s["sample_id"]]}
        joined.append(merged)

# Group by cohort
groups = defaultdict(list)
for row in joined:
    groups[row["cohort"]].append(row)

# Summarize
summary = []
for cohort, rows in groups.items():
    summary.append({
        "cohort": cohort,
        "count": len(rows),
        "mean_depth": mean(r["depth"] for r in rows),
        "mean_quality": mean(r["quality"] for r in rows),
        "total_reads": sum(r["read_count"] for r in rows),
    })

summary.sort(key=lambda x: x["mean_depth"], reverse=True)

print("Cohort Summary:")
for row in summary:
    print(f"  {row['cohort']}: n={row['count']}, depth={row['mean_depth']:.1f}, "
          f"qual={row['mean_quality']:.1f}, reads={row['total_reads']}")

# High-quality filter
hq = [r for r in joined if r["quality"] >= 30 and r["depth"] >= 20]
print(f"\nHigh-quality samples: {len(hq)} / {len(joined)}")
